### PR TITLE
[ATEN->C10] Migrate return type void to () for native functions.

### PR DESCRIPTION
**Pull request by xcheng16**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27192 [ATEN->C10] Migrate return type void to () for native functions.**

Differential Revision: [D17565528](https://our.internmc.facebook.com/intern/diff/D17565528/)

### CODE REVIEW

File: aten/src/ATen/function_wrapper.py [hunk 1]
```diff
@@ -1,6 +1,8 @@
 # HEY! Trying to understand what this file does?  Read
 # "what has to be done to add a Operation ..." first!
 
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: I know you're just trying to fix fbcode lint here, but when you do this please actually check that this change is semantics preserving. Often it's not!

Comment (ezyang) [reply aten/src/ATen/function_wrapper.py]: I did a quick audit for divisions and it looks like we're ok.

File: aten/src/ATen/function_wrapper.py [hunk 2]
```diff
@@ -887,7 +889,10 @@ def get_return_types(option):
 
     def format_return_type(return_types):
         # type: (List[ReturnType]) -> str
-        if len(return_types) == 1:
+        len_return_types = len(return_types)
+        if len_return_types == 0:
+            return "void"
+        if len_return_types == 1:
```
Comment (ezyang) [inline aten/src/ATen/function_wrapper.py]: nit (this means no change is needed here, this is just some tips for the next time you write some code): len should be O(1) on lists in Python, so you don't have to assign it to a variable. Also, intent is clearer if you use `elif` for the second if statement.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -37,10 +37,11 @@
   use_c10_dispatcher: full
   variants: function
 
-- func: backward(Tensor self, Tensor? gradient=None, bool keep_graph=False, bool create_graph=False) -> void
+- func: backward(Tensor self, Tensor? gradient=None, bool keep_graph=False, bool create_graph=False) -> ()
   variants: method
 
-- func: set_data(Tensor(a!) self, Tensor new_data) -> void
+- func: set_data(Tensor(a!) self, Tensor new_data) -> Tensor(a!)
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml]: @yf225 @pbelevich This makes `set_data` support method chaining. Seem good to you? (I also kind of wonder why we called this `set_data` and not `data_`.)

File: tools/autograd/gen_variable_type.py [hunk 4]
```diff
@@ -509,7 +509,8 @@ def emit_body(declaration):
     inplace = declaration['inplace']
     is_out_fn = name.endswith('_out')
     modifies_arguments = inplace or is_out_fn
-    returns_void = len(returns) == 1 and returns[0]['type'] == 'void'
+    len_returns = len(returns)
```
Comment (ezyang) [inline tools/autograd/gen_variable_type.py]: nit: this variable assignment is pointless, it's single use :)

**Review by ezyang — APPROVED** [review]
Looks good, thank you!

**Comment by ezyang** [comment]
your build failures are due to merge conflict on master. Rebase and then rebuild with `GEN_TO_SOURCE` to get updated generated files

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (yf225) [reply aten/src/ATen/native/native_functions.yaml]: Yes it sounds great. We were following the naming for the `Variable::set_data` method, but maybe we should change it to `data_` instead.

**Pull request closed without merging**